### ISSUE TITLE
Optimize build times

### ISSUE THREAD

**Issue by kvark** [amethyst/rendy#203]
Removing `failure` and `derivative` should be a solid step in this direction. See https://github.com/gfx-rs/gfx/pull/2970 and also https://github.com/amethyst/rendy/pull/198

**Comment by Veykril** [comment]
As already said in #210, removing `derivative` does not seem to have any major effect in build times. I've just tested out how `failure` affects the build times by removing it and replacing the `failure::Errors` with basic appropriate structs/enums. Given how sparse it is actually used it didnt seem to have any major effect on build times either, resulting in `24.29s` build time for just the rendy crates and `1m 04s` for a full build, compared to the previous `24.53s` and `1m 08s`. The 4 second cut off is most likely due to `failure` not being a dependency in the whole project anymore thanks to the removal. (All the timings were done with possible background noise though, which is why i had one rendy crates-only build with just 20s).

Nevertheless I'd say to remove `failure` given that it seems to have been just used as a convenient error wrapping for strings. One or two rendy crates even have it as a left over dependency without even using it in the source like `rendy-frame` for example.

So if the consensus is to remove `failure` I'll be happy to clean up my test changes and make a PR for it.

**Comment by kvark** [comment]
#198 has largely removed `failure`, it's just a few places left before it can be expelled from the Cargo toml. That, I think, explains why you aren't seeing much gains from finishing this work.
Overall, it would be great to see your PR doing this. Derivative can stay, for now :)

**Comment by malobre** [comment]
Compile time report from [cargo-bloat](https://crates.io/crates/cargo-bloat):
```
$ cargo bloat --release --time -j 1
Compiling ...

  Time Crate
53.56s serde_derive
36.94s winit
36.69s syn
34.63s syn
32.71s wayland_protocols
31.66s derivative
27.23s smithay_client_toolkit
25.83s syn
18.27s synstructure
17.74s wayland_client
17.18s wayland_sys
15.61s gfx_hal
13.25s serde
12.90s x11_dl
12.37s wayland_scanner
12.17s nix
12.12s rendy_graph
11.86s cc
10.43s rendy_command
 9.96s rustc_serialize
 9.50s rendy_shader
 8.53s proc_macro2
 8.52s proc_macro2
 8.40s proc_macro2
 8.28s xml
 8.03s serde_json
 7.99s rendy_factory
 7.91s rendy_memory
 7.57s andrew
 7.43s rendy_util
 7.18s failure_derive
 7.08s spirv_reflect
 6.77s rendy_chain
 6.06s num_bigint
 5.09s mio
 4.93s thread_profiler
 4.51s pkg_config
 4.32s rendy_resource
 4.18s cmake
 3.75s spirv_headers
 3.57s jobserver
 3.49s parking_lot
 3.34s num_rational
 3.26s rand
 3.01s rusttype
 2.94s shaderc_sys
 2.92s num_derive
 2.84s quote
 2.77s num_traits
 2.74s time
 2.68s walkdir
 2.62s xdg
 2.55s semver_parser
 2.51s rendy_mesh
 2.43s libc
 2.41s crossbeam_channel
 2.41s rendy_wsi
 2.41s rendy_descriptor
 2.40s backtrace
 2.31s crossbeam_utils
 2.31s byteorder
 2.27s rustc_demangle
 2.10s quote
 2.10s wayland_commons
 2.09s semver
 2.07s autocfg
 2.02s rendy_texture
 1.97s arrayvec
 1.86s libloading
 1.82s backtrace_sys
 1.80s parking_lot_core
 1.80s shaderc
 1.77s num_cpus
 1.67s num_integer
 1.66s rustc_version
 1.65s net2
 1.64s ryu
 1.63s colorful
 1.50s quote
 1.46s hibitset
 1.36s stb_truetype
 1.23s calloop
 1.21s log
 1.12s mio_extras
 1.07s rendy_frame
 1.06s bitflags
 1.03s maybe_uninit
 0.95s failure
 0.90s num_complex
 0.86s percent_encoding
 0.61s num_iter
 0.57s memmap
 0.55s smallvec
 0.49s lock_api
 0.41s same_file
 0.37s ordered_float
 0.33s relevant
 0.31s approx
 0.29s either
 0.29s iovec
 0.26s itoa
 0.25s line_drawing
 0.23s rendy
 0.21s slab
 0.20s unicode_xid
 0.20s unicode_xid
 0.18s raw_window_handle
 0.15s atom
 0.15s void
 0.14s lazycell
 0.12s lazy_static
 0.12s downcast_rs
 0.11s num
 0.11s dlib
 0.11s scopeguard
 0.10s rusttype
 0.09s nodrop
 0.09s cfg_if
```

**Comment by kvark** [comment]
The low hanging fruit here is deriving `Deserialize` behind a `deserialize` feature that's not enabled by default. See how [webrender does it](https://github.com/servo/webrender/blob/b4f4831359d02604ec8d7abf0fade8bf98aad87d/webrender/src/clip.rs#L109-L117) to a large extent. Note: `Deserialize` is known to take significantly more time than `Serialize`.

**Comment by malobre** [comment]
Here is a dependency graph:
![deps](https://user-images.githubusercontent.com/1368457/65449308-ffa53f80-de3a-11e9-8c44-5c66e4ee123c.png)

I know, it's hard on the eyes ahah.

**Comment by omni-viral** [comment]
Am I correct that removing `syn` duplcation would also save us ~30s?

**Comment by malobre** [comment]
Indeed!
`syn` is built 3 times: 
* 36.69s
* 34.64s
* 25.83s

**Comment by malobre** [comment]
`syn v0.14` -> `palette_derive` -> `palette` -> `rendy-texture`
`syn v0.12` -> `num_derive v0.1` -> `spriv_headers` -> `spriv_reflect` -> `rendy-shader`
`syn v0.15` -> `serde` and `derivative`
`syn v1.05` -> a bunch of crates

**Comment by kvark** [comment]
@malobre @omni-viral as I'm trying to vendor `rendy-memory` and `rendy-descriptor` crates as 3rd party dependencies in Firefox, the extra code of Rendy dependencies is hard to justify.

Please consider removing `derivative`, `failure`, and any other dependencies that aren't required (or make them optional). This is an issue today for us.

**Comment by malobre** [comment]
- `failure` was removed in #211
- `derivative` has also been removed in #210 (__but not merged__, see PR for discussion)

**Comment by malobre** [comment]
I have time to work on that, but I have one question.
Should I remove the `serde` feature and create two new one (one for serialize, one for deserialize) or should I keep the `serde` feature which would enable both new features (for back-compatibility) ?

**Comment by kvark** [comment]
@malobre good question!
First of all, `Deserialize` is significantly slower to generate than `Serialize`. That's the reason https://github.com/servo/webrender/ has separate features ("capture" vs "replay") for deriving these traits.

Rendy though would mostly care about what gfx-rs derives today, and that's just controlled by `serde` optional dependency. Maybe in the future we'll split it like WebRender does, but it hasn't shown up as a problem yet.

When you measured the times, why did `serde` even show up? Is it accidentally enabled by default somewhere?

Another question - was the `syn` duplication problem resolved?

**Comment by malobre** [comment]
Oh I see what the problem is, the timings were with __all features enabled__. So the `serde` flag is probably sufficent.

As for the `syn` duplication this is still not resolved.
To resolve:
- [ ] `spirv_headers` needs to update `num_derive ^0.1` to `num_derive ^0.3` (which depends on syn)
- [ ] `palette_derive` needs to update `syn ^0.14` to `syn ^1`
- [ ] Either remove `derivative` (#210) or update it's `syn ^0.15.10` dependency to `syn ^1`

After all these changes we will only be building `syn ^1` 1 time.

**Comment by kvark** [comment]
So removing the `derivative` is still an option.

**Comment by malobre** [comment]
@Frizi how do you feel about merging #210 ?
If we keep the `derivative` dependency I will make a PR on their repo to update to `syn ^1`

**Comment by Veykril** [comment]
Removing `derivative` from just the two crates mentioned(`rendy-memory` and `rendy-descriptor`) might also be an option for the time being, since there are only 3 derives that are made through `derivative` for those crates. So the maintenance reason wouldn't apply too much to those two crates https://github.com/amethyst/rendy/pull/210#issuecomment-534108290.

**Comment by omni-viral** [comment]
TBH I'm for getting rid of `derivative` because it's slow and not all features described in their doc works.

**Comment by Frizi** [comment]
> getting rid of derivative because it's slow
Is it though? From what we've seen before in #210, it isn't really that significant after all.

This might be changed if we get rid of extra copy of `syn` by removing it.

@malobre I don't know tbh. I kinda start to think that it might be good idea to get rid of it, just to make it a bit easier on firefox dev team and alike. All the little libraries add up. My conern is that applying this logic to many other dependencies is dangerous. The concept of sharing code is useful after all, it can actually bring improvements to the build times. We just need to keep things up to date.


Another kinda separate point is, I think we should focus a little bit more on making the rendy's code alone faster to compile. I'd love to have some kind of tools that tell me where the compiler spends most time, maybe we can help it a bit. Also a big factor for sure is the proliferation of generics that gets monomorphized in the user's code. It's really hard to get rid of some of them (notably backend), but maybe we can at least monomorphize them in rendy's code already.

**Comment by malobre** [comment]
I agree about the dependency situation, this is all about finding an equilibrium. The `syn` situation will take some time to resolve.

About rendy code, I'll try to establish what takes time to compile and make a little summary so we have an idea where optimization would be nice.

**Comment by malobre** [comment]
`$ cargo +nightly build -Z timings` yields some amazing graphs
![Screenshot_2019-10-31 Cargo Build Timings — rendy 0 5 0](https://user-images.githubusercontent.com/1368457/67941966-0744ca80-fbd7-11e9-92b1-45632793984c.png)

**Comment by omni-viral** [comment]
Whoa, why `rendy-core` is so slow? It's mostly a bunch of exported macros.

**Comment by malobre** [comment]
One indirect `syn` dependency was removed. (https://github.com/gwihlidal/spirv-reflect-rs/issues/10)

**Comment by malobre** [comment]
@omni-viral could you `cargo clean && cargo +nightly build -Z timings` and upload the resulting graphs ? To compare with my results.

**Comment by omni-viral** [comment]
Current nightly compiler ICE on rendy without backend features.

**Comment by malobre** [comment]
https://github.com/Ogeon/palette/issues/145 has been closed, and `palette_derive^0.5` has been published on crates.io, it now requires `syn^1` (same as derivative).

We should only be building `syn` one time now.

**Comment by ezpuzz** [comment]
implemented dependabot to keep dependencies up to date, individual issues can be raised about time consuming dependencies, duplicates in tree, etc.

palette upgrade has PR already

would like to automate the timings result being posted to PRs, will file issue
